### ISSUE TITLE
[ENH] Autoregressive Wrapper

### ISSUE THREAD

**Issue by benHeid** [sktime/sktime#6802]
**Is your feature request related to a problem? Please describe.**
Some of the forecaster are only capable of doing fixed length predictions. E.g., KanForecaster most reduction based forecaster, ... 
This might be disadvantageous, if the forecaster takes a lot longer to train if the number of outputs increases. Furthermore, it also leads to problems if such a forecaster is used in some probabilistic wrappers (e.g., ConformalIntervals, since the FH varies there for the different forecasts).

However, with an autoregressive wrapper, this problem could be solved. 

## Possible Solution 1 (GlobalForecasting-based Solution)
Forecaster, that should be wrappable by such a wrapper need to inherit from `GlobalBaseForecaster`. In this case, the wrapper could take care of handling the autoregressive behavior:

**Implementation Code** (more some kind of pseudo-code:

```python

class AutoregressiveWrapper(BaseForecaster/BaseGlobalForecaster):

    def __init__(self, forecaster, aggregate_method):
        self.forecaster = forecaster
        self.aggregate_method = aggregate_method


    def _fit(self, ....):
        self.forecaster.fit(...)

    def _predict(self, .....):
        max_fh = .... # The last point that should be predicted
        index_that_should_be_predicted = .... # The values that need to be predicted

        _y = pd.concat([y, pd.DataFrame(index=range(index_that_should_be_predicted))]
        predictions = []
        for i, idx in enumerate(index_that_should_be_predicted[:-max(self.forecaster.fh.to_relative())]
            __y, __X = self.get_input_values(_y, X, forecaster.fh) # assume that fh is set always
            predictions.append(self.forecaster(y=__y, X=__X)
            self._append_new_value(__y, self.aggregate_method(pd.concat(predictions)).loc[idx])
       return self.aggregate_method(pd.concat(predictions))         
```

**User Code**

```python

forecaster = SomeGlobalForecaseter(...)
ar_forecaster = AutoregressiveWrapper(forecaster, np.mean)
ar_forecaster.fit(...)
ar_forecaster.predict(....)
```

## Possible Solution 2 (internal solution):

We could try to add a method to BaseForecaster, or some new method, that is introducing a method enabling autoregressive behaviour. 
I would assume, that this could be controlled via a tag (e.g. `enable_ar`). If this is enabled, it is not anymore forced that the (`fh` from training is the same as the `fh` for the predict) and in `predict` the private `_predict` would be called as much as it would be needed to forecast all requested values (this would probably look similar as the code above, with some data updates instead of using `_append_new_value`)

**Comment by fkiraly** [comment]
I prefer solution 1, since the added logic seems extraneous to the base interface. From a design perspective, we should avoid coupling any modelling logic to the top level base class (the broadcasting is already in a grey area ...)

**Comment by geetu040** [comment]
I would like to work on this, please assign me this issue.

**Comment by geetu040** [comment]
@fkiraly @benHeid can you please review this design below? emphasis is on `horizon_length` as it was not in the initial design proposed by @benHeid 
```py
class AutoRegressiveWrapper(_BaseGlobalForecaster):
    def __init__(self, forecaster, aggregate_method, horizon_length):
        pass

    def _fit(self, y, X=None, fh=None):
        _fh = range(1, self.horizon_length+1)
        self.forecaster.fit(y=y, X=X, fh=_fh)

    def _predict(self, fh, X=None, y=None):
        max_fh = ... # find max from `fh`
        # regressive forecasting till max_fh

forecaster = NaiveForecaster()
wrapper = AutoRegressiveWrapper(
    forecaster=forecaster,
    horizon_length=4,
)

wrapper.fit(y=y_train)
# predict till this length regressively
wrapper.predict(fh=[10, 11, 12])
# global predict this length regressively
wrapper.predict(y=y_test, fh=[1, 2, 3, 4, 5, 6, 7])
```

**Comment by benHeid** [comment]
I think this make sense.

In and additional step, we might also think about if it makes sense to also enable probabilistic forecasting if the underlying forecaster is able to do probabilistic forecasting

**Comment by geetu040** [comment]
@benHeid I have a few quesitons regarding the design

1. The wrapper has to be of strictly **global forecasting nature**, as it uses the future values as history predicts upon that. This in something only global forecasters can do. So am I right to inherit the Global Forecasting base class and design this to be specifically used by global forecasters only?
2. do we ignore the **exogenous data**? I am not sure how that works with global forecasting, but I'll study that if needed.
3. what is meant by the **aggregation method**? how does that work? I am simply concatenating the new forecasted values into the history. where is aggregation used? I cannot understand from the first example.
4. the only global forecaster available at the moment is from pytorch forecasting, so I'll add that in the **test_params** for this wrapper, which means if `pytorch-forecasting` package is not installed, there will be an empty test_params, will that be okay?

**Comment by benHeid** [comment]
> 1. The wrapper has to be of strictly **global forecasting nature**, as it uses the future values as history predicts upon that. This in something only global forecasters can do. So am I right to inherit the Global Forecasting base class and design this to be specifically used by global forecasters only?


Yes, it should inherit from global and all passed forecasters are global forecaster

> 2. do we ignore the **exogenous data**? I am not sure how that works with global forecasting, but I'll study that if needed.

No, I would say exogenous should be supported. You need to pass exogenous values that match the forecasting horizon. 

> 3. what is meant by the **aggregation method**? how does that work? I am simply concatenating the new forecasted values into the history. where is aggregation used? I cannot understand from the first example.

No,depending on the prediction length multiple predictions per timestamp might exist. Thus aggregation method should be mean or median or something else. 


> 4. the only global forecaster available at the moment is from pytorch forecasting, so I'll add that in the **test_params** for this wrapper, which means if `pytorch-forecasting` package is not installed, there will be an empty test_params, will that be okay?

Yes but perhaps we should add support to a naive forecaster to speed up tests. 

> 5. which category this forecaster suits in the [forecasting.rst](https://github.com/sktime/sktime/blob/main/docs/source/api_reference/forecasting.rst)? I think it should be in "Direct and recursive - sktime native 1st/2nd generation"

I am not sure what you are referring to.

**Comment by geetu040** [comment]
> > 3. what is meant by the **aggregation method**? how does that work? I am simply concatenating the new forecasted values into the history. where is aggregation used? I cannot understand from the first example.
> 
> No,depending on the prediction length multiple predictions per timestamp might exist. Thus aggregation method should be mean or median or something else.

suppose we want to predict for `fh=[1, 2, 3, 4, 5, 6]` and `prediction_length=3` specified by the user.
forecaster uses history of timestamps `[t0, t1, t2, ..., tn]` and predicts values at timestamps `tn+1, tn+2, tn+3`. now these values are given as history for next values, i.e a history of timestamps `t3, t4, t5, ..., tn+1, tn+2, tn+3` predicts next values at timestamps `tn+4, tn+5, tn+6`.
`t0` here refers to a value in times series at index `t0`, the value may be 12, 20, 0.6 or anything.
@benHeid  how is aggregation applied here?
are you suggesting that we keep a sliding window of 1 instead of prediction_length when preparing next history for the forecaster?

@fkiraly  for your context this wrapper extends a given forecasters for longer predictions by regressive forecasting. the given forecaster may be trained to predict only a fixed length, but the wrapper extends it.

**Comment by fkiraly** [comment]
I have to admit I do not exactly get the algorithm to be implemented, based on the above - some pseudocode with minimal math or array notation would help me

**Comment by fkiraly** [comment]
oh, I see - the typical and clear case being, where `horizon_length=1`, but it being unclear what to do when it is sth else?

**Comment by geetu040** [comment]
> oh, I see - the typical and clear case being, where `horizon_length=1`, but it being unclear what to do when it is sth else?

I am a little unclear here myself, I would say lets wait for @benHeid reply to this https://github.com/sktime/sktime/issues/6802#issuecomment-2256506898. As I need to grasp his perspective on how and what aggregation is doing.

**Comment by benHeid** [comment]
I agree that it is not clear on how to aggregate these values. Thus, I would let the user decide what to choose and proposing as default the mean. You can represent this internally as a matrix:


In the following example, the forecaster would always predict the next two values
```
....
Prediction 3 ..xx.........
Prediction 2 .xx..........
Prediction 1 xx ...........
FH ----------12345------------------> 

```

In this case, for FH 1, only the context would be used, for FH 2 the context and the aggregation of the first column, for FH 3 the aggregation of the second column, etc.

The aggregation itself would be then a parameter of the AutoregressiveWrapper: This could be mean, median, or some other custom function specified by the user. Thus, I would propose that the `type(agggregate_method)==Callable`.
